keps/sig-node/4815-dra-partitionable-devices/README.md

```diff
@@ -0,0 +1,2140 @@
+	// shared attributes and capacities that an actual device can "include"
+	// to extend the set of attributes and capacities it already defines.
+	//
+	// The main purposes of these mixins is to reduce the memory footprint
+	// of devices since they can reference the mixins provided here rather
+	// than duplicate them.
+	//
+	// The total number of mixins, basic devices, and composite devices must be
+	// less than 128.
+	//
+	// +optional
+	// +listType=atomic
+	DeviceMixins []DeviceMixin `json:"deviceMixins,omitempty"`
+}
+
+// DeviceMixin defines a specific device mixin for each device type.
+// Besides the name, exactly one field must be set.
+type DeviceMixin struct {
+	// Name is a unique identifier among all mixins managed by the driver
+	// in the pool. It must be a DNS label.
+	//
+	// +required
+	Name string `json:"name"`
+
+	// Composite defines a mixin usable by a composite device.
+	//
+	// +optional
+	// +oneOf=deviceMixinType
+	Composite *CompositeDeviceMixin `json:"composite,omitempty"`
+}
+
+// CompositeDeviceMixin defines a mixin that a composite device can include.
+type CompositeDeviceMixin struct {
+	// Attributes defines the set of attributes for this mixin.
+	// The name of each attribute must be unique in that set.
+	//
+	// To ensure this uniqueness, attributes defined by the vendor
+	// must be listed without the driver name as domain prefix in
+	// their name. All others must be listed with their domain prefix.
+	//
+	// Conflicting attributes from those provided via other mixins are
+	// overwritten by the ones provided here.
+	//
+	// The maximum number of attributes and capacities combined is 32.
+	//
+	// +optional
+	Attributes map[QualifiedName]DeviceAttribute `json:"attributes,omitempty"`
+
+	// Capacity defines the set of capacities for this mixin.
+	// The name of each capacity must be unique in that set.
+	//
+	// To ensure this uniqueness, capacities defined by the vendor
+	// must be listed without the driver name as domain prefix in
+	// their name. All others must be listed with their domain prefix.
+	//
+	// Conflicting capacities from those provided via other mixins are
+	// overwritten by the ones provided here.
+	//
+	// The maximum number of attributes and capacities combined is 32.
+	//
+	// +optional
+	Capacity map[QualifiedName]DeviceCapacity `json:"capacity,omitempty"`
+}
+
+// Device represents one individual hardware instance that can be selected based
+// on its attributes. Besides the name, exactly one field must be set.
+// +k8s:deepcopy-gen=true
+type Device struct {
+	// Name is unique identifier among all devices managed by
+	// the driver in the pool. It must be a DNS label.
+	//
+	// +required
+	Name string `json:"name"`
+
+	// Basic defines one device instance.
+	//
+	// +optional
+	// +oneOf=deviceType
+	Basic *BasicDevice
+
+	// Composite defines one composite device instance.
+	//
+	// +optional
+	// +oneOf=deviceType
+	Composite *CompositeDevice `json:"composite,omitempty"`
```

Comment:
Yeah. I originally called it `PartitionableDevice` but that put the emphasis on the the `consumeCapacityFrom` field and ignored the new `mixin` field. 

I'd love it if we could just add both of these new fields to the existing `BasicDevice` type (since they are strictly extensions), but as @pohly keeps pointing out (because I keep forgetting) we can't extend the API in this way because it breaks downgrades. An old scheduler will not recognize the new fields and attempt to allocate a device based solely on the fields that it DOES know about. For mixins / attributes this is not *too* bad because selection would simply fail if the scheduler was not able to see the additional attributes that a mixin provides.  Not honoring `consumesCapacityFrom` properly would be unacceptable though.